content/05.main-text.md

```diff
@@ -749,10 +767,68 @@ Producing results for these queries relies on internal calls to the PostgreSQL r
+
+### Webapp and Frontend
+
+We developed the [connectivity search app](https://het.io/search), as well as the related [rephetio](https://het.io/repurpose/) and [disease-associated genes](https://het.io/disease-genes/) apps, as single-page, standalone applications using React and associated tools.
+The source code for the connectivity search app is available in the [`connectivity-search-frontend`](https://github.com/greenelab/connectivity-search-frontend) repository, and the source code for the related apps are available in the [hetio GitHub organization](https://github.com/hetio).
+
+Since the rest of the overarching Hetio website was mostly non-interactive content, it was appropriate to construct the bulk of the website in simple static formats like Markdown and HTML using Jekyll, and leave React for implementing the sections of the site that required more complex behavior.
+
+We used React's own `create-react-app` command line tool to generate a boilerplate for the app.
+This greatly simplified setting up and maintaining the app's testing and building pipelines, bypassing time-consuming configuration of things like Webpack and linters.
+Some configuration of was necessary to produce non-hashed, consistently named output files like `index.js` that could be easily and reliably referenced by and embedded into the Hetio Jekyll website.
+
+For authoring components, we used React's traditional class syntax.
+At the time of authoring the app, React Hooks were still nascent, thus the simpler and less-verbose functional syntax was not viable.
+
+We created a custom package of React components and utility functions that could be shared across the multiple interactive apps on the website.
+The package consists of interface "components" (building blocks) like buttons, sortable/searchable/paginated tables, etc., as well as utility functions for formatting data, debugging, etc.
+Each of the interactive apps import this package to reduce code repetition, and to enforce a consistent style and behavior across the website.
+
+For managing state in the connectivity search app, we used the Redux library.
+In general, Redux is a well-accepted approach to managing complex state.
+To be more explicit, Redux was chosen over vanilla React or other state management libraries for a few reasons:
+
+1) The state in this app was very "global", meaning most of it was needed by a lot of different parts of the app.
+Redux provides a convenient global "store" of state that is easily accessible to any component, avoiding the "prop-drilling" phenomenon.
+2) The structure of the state is nested and complex.
+Redux's "reducer" approach make it cleaner to modify this type of data.
+3) Redux's approach to immutable state that is updated by actions and pure functions makes the application easier to debug.
+It is easy to get a clear timeline of how and when the state changed, and what changed it. 
+
+To create the graph visualization at the bottom of the app, the D3 library was used.
```

Comment:
Should we link D3 to https://d3js.org/ upon first mention? Same with React and other software mentioned?

I see that I didn't do this in the database section, so not essential.

Reply:
Yeah I just went based off of your pattern there. There are several libraries and concepts mentioned and it might be a bit much to link to all of them. Up to you, I can easily include if you want.